vpr/src/route/connection_router.h

```diff
@@ -38,20 +39,25 @@ struct t_conn_cost_params {
 // rr_node_route_inf.  The routed path can be found by tracing from the sink
 // node (which is returned) through the rr_node_route_inf.  See
 // update_traceback as an example of this tracing.
+template<typename HeapImplementation>
```

Comment:
I think its worth discussing whether using a template (i.e. compile time polymorphism) or relying on the virtual functions of the HeapInterface (i.e. run-time polymorphism) is the right approach here.

I quite like the idea of having a command-line option to control which heap is in use (e.g. `--router_heap { binary | bucket }`), as it makes it easier to fall-back to the binary heap if necessary (e.g. if due to bugs or other poor behavior).

Compile time polymorphism is theoretically more run-time efficient, but I expect a the virtual function call to easily predictable in this case (it would always go the same way during a run of VPR). So I think this is something worth considering, and most importantly measuring.

Reply:
We can still have a command-line option to switch between heaps with the template, it just costs the size of the code increase.  I expect template performance to be significantly faster than run-time polymorphism.

Reply:
> Compile time polymorphism is theoretically more run-time efficient, but I expect a the virtual function call to easily predictable in this case (it would always go the same way during a run of VPR). So I think this is something worth considering, and most importantly measuring.

The cost is the inability to inline across the vtable call.  The overhead of pushing and popping the stack for that call is likely significant.  I agree measurements are required.

Reply:
> We can still have a command-line option to switch between heaps with the template, it just costs the size of the code increase.

Yep, that may make the most sense. Are you envisioning a run-time option to choose between ConnectionRouter<BinaryHeap> or ConnectionRouter<BucketHeap>? If we can do that reasonably cleanly I'd be happy with it.

I'm not really concerned about the mechanism (template vs virtual), just that it would be good to be able to switch the heap with a command-line option.

Reply:
> I'm not really concerned about the mechanism (template vs virtual), just that it would be good to be able to switch the heap with a command-line option.

The PR has been updated with a command line flag already (~2 minutes ago), so take a look.